File: packages/python/src/quanttide_agent/message.py
"""Message and response models for conversation."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel

from .cost import Usage
from .tool import ToolCall


class Message(BaseModel):
    """A message in a conversation.

    >>> m = Message(role="user", content="hello")
    >>> m.to_dict()
    {'role': 'user', 'content': 'hello'}
    """

    role: Literal["system", "user", "assistant", "tool"]
    content: str
    tool_call_id: str | None = None

    def to_dict(self) -> dict:
        d: dict[str, Any] = {"role": self.role, "content": self.content}
        if self.tool_call_id:
            d["tool_call_id"] = self.tool_call_id
        return d


class ChatResponse(BaseModel):
    """Response from an LLM chat call.

    >>> resp = ChatResponse(content="Hello!", model="deepseek-v4-pro")
    >>> resp.content
    'Hello!'
    """

    content: str
    model: str
    finish_reason: str = "stop"
    reasoning_content: str | None = None
    tool_calls: list[ToolCall] | None = None
    usage: Usage | None = None

File: packages/rust/src/message.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::cost::Usage;
use crate::tool::ToolCall;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub role: String,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
}

impl Message {
    pub fn new(role: &str, content: &str) -> Self {
        Self {
            role: role.to_string(),
            content: content.to_string(),
            tool_call_id: None,
        }
    }

    pub fn to_dict(&self) -> Value {
        let mut map = serde_json::Map::new();
        map.insert("role".to_string(), Value::String(self.role.clone()));
        map.insert("content".to_string(), Value::String(self.content.clone()));
        if let Some(ref id) = self.tool_call_id {
            map.insert("tool_call_id".to_string(), Value::String(id.clone()));
        }
        Value::Object(map)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatResponse {
    pub content: String,
    pub model: String,
    #[serde(default = "default_finish_reason")]
    pub finish_reason: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reasoning_content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<Vec<ToolCall>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub usage: Option<Usage>,
}

fn default_finish_reason() -> String {
    "stop".to_string()
}
